components/search/backend/src/routes/kotlin/Validation.kt

```diff
@@ -0,0 +1,122 @@
+import org.eclipse.apoapsis.ortserver.components.authorization.rights.ProductPermission
+import org.eclipse.apoapsis.ortserver.components.authorization.rights.RepositoryPermission
+import org.eclipse.apoapsis.ortserver.components.authorization.routes.AuthorizationChecker
+import org.eclipse.apoapsis.ortserver.components.authorization.service.AuthorizationService
+import org.eclipse.apoapsis.ortserver.model.CompoundHierarchyId
+import org.eclipse.apoapsis.ortserver.model.HierarchyId
+import org.eclipse.apoapsis.ortserver.model.OrganizationId
+import org.eclipse.apoapsis.ortserver.model.ProductId
+import org.eclipse.apoapsis.ortserver.model.RepositoryId
+
+/**
+ * Parse the scope from query parameters. Returns the scope [HierarchyId] or null if no scope is provided.
+ * Returns an error if multiple scope parameters are provided.
+ *
+ * This function enforces that only one scope parameter (organizationId, productId, or repositoryId) can be
+ * provided at a time to prevent authorization bypass attacks where a user could specify an organization they
+ * have access to along with arbitrary product/repository IDs.
+ */
+fun ApplicationCall.parseScope(): Result<HierarchyId?> {
+    val organizationIdParam = request.queryParameters["organizationId"]?.toLongOrNull()
+    val productIdParam = request.queryParameters["productId"]?.toLongOrNull()
+    val repositoryIdParam = request.queryParameters["repositoryId"]?.toLongOrNull()
+
+    val scopeCount = listOfNotNull(organizationIdParam, productIdParam, repositoryIdParam).size
+
+    if (scopeCount > 1) {
+        return Result.failure(
+            IllegalArgumentException(
+                "Only one scope parameter (organizationId, productId, or repositoryId) is allowed."
+            )
+        )
+    }
+
+    val scope = when {
+        repositoryIdParam != null -> RepositoryId(repositoryIdParam)
+        productIdParam != null -> ProductId(productIdParam)
+        organizationIdParam != null -> OrganizationId(organizationIdParam)
+        else -> null
+    }
+
+    return Result.success(scope)
+}
+
+/**
+ * Create an [AuthorizationChecker] that checks permissions dynamically based on the scope query parameters.
+ * - If no scope is provided, requires superuser permission.
+ * - If organizationId is provided, requires organization READ permission.
+ * - If productId is provided, requires product READ permission.
+ * - If repositoryId is provided, requires repository READ permission.
+ *
+ * Note: Only one scope parameter is allowed at a time to prevent authorization bypass attacks.
+ * Use [parseScope] to extract and validate the scope from request parameters.
+ */
+fun requireScopedReadPermission(): AuthorizationChecker =
+    object : AuthorizationChecker {
+        override suspend fun loadEffectiveRole(
+            service: AuthorizationService,
+            userId: String,
+            call: ApplicationCall
+        ): EffectiveRole? {
+            val scope = call.parseScope().getOrNull()
```

Comment:
IIUC, if `parseScope` returns a failed result, then superuser rights are required. Wouldn't it be better to rethrow the exeption (or some other exception), so that the endpoint can return a 400 response in this case?